[3.x] Reset joy axes and buttons after joy disconnection

orlambda opened the pull request:
Fixes #105047

orlambda committed:
Reset joy axes and buttons after joy disconnection

lawnjelly reviewed (dismissed):
I'm not super expert in the joystick handling but I think this looks correct.

Would be nice to get some input ( :grin: ) from a more regular input contributor to check so I'll post on rocketchat before going further.

MJacred reviewed (changes requested):
I probably don't count as a more regular input contributor, but here I go:

The approach is a hack-fix, as it basically sends another event for a device that has been disconnected.

this results currently in…
* one InputEventJoypadMotion event in `_input` (with axis == `0.0`) for _each_ axis I used before disconnecting. It makes no difference whether I hold the stick on disconnect or not
* the first `_process()` call after disconnect outputs the axis value when using `Input.get_action_strength` , not `0.0`

lawnjelly commented on main/input_default.cpp:
```diff
@@ -285,29 +285,32 @@ void InputDefault::joy_connection_changed(int p_idx, bool p_connected, String p_
-				set_joy_axis(p_idx, i, 0.0f);
+		for (int i = JOY_BUTTON_0; i < JOY_BUTTON_MAX; i++) {
+			if (is_joy_button_pressed(p_idx, i)) {
+				_button_event(p_idx, i, 0.0f);
```
Good spot.

lawnjelly commented:
Just posting this for the author so they can see as discussed in input on rocketchat:

MJacred
2:27 PM
reviewed the code and tested it under various circumstances (and checked against 3.6)

lawnjelly
3:08 PM
The events on disconnect I'm not so sure is a problem (this seems logical to receive cancellation of buttons and axis in user code?), but you could be right about the first _process() call outputting the axis value rather than 0.0 that doesn't sound right... 🤔
Essentially if a device is disconnected, is there a problem with sending events that effectively cancel anything that is pressed?
That means user code can respond to it how you would expect I would think? 🤔

MJacred
3:46 PM
I don't think it's an issue if you get a proper "cancel" event.
both bugs address nonsense behaviour:
_process() gives a value which should have been reset for a device which is not addressable anymore
_input() triggers events on disconnect for actions that happened somewhere in the past. and triggers them again, if you re-connect your joypad and disconnect again w/o touching an axis

lawnjelly
3:49 PM
Is there a problem caused by receiving event from disconnected device? Will it crash, or is it possibility of connecting another device to the same ID or something?
In most cases you would have thought user code would react ok to a 0.0 event (or press being false)... 🤔
It might be possible to log on the joystick buttons whether they have been ever pressed while connected, but is it actually worth it I'm wondering.
I.e. is this a hypothetical problem rather than real world, and could we fix that independently if it was reported as a bug later?
I do take the point that it is not perfect, but we also see a lot of "perfect is the enemy of good enough" so we don't want to lose out on a fix if it makes the situation better than the current.
I'm not an input guy, but that's my thoughts for release management.

MJacred
3:54 PM
regarding the _input(): I reckon a _joy_axis.clear(); is missing, i.e. InputDefault::release_pressed_events() is not used. that's why it re-triggers the same events over and over on disconnect

lawnjelly
3:59 PM
So maybe something like:
if (_joy_axis.has(c) && _joy_axis[c] != 0)
{
_axis_event(p_idx, i, 0.0f);
}

or something like that (pseudocode)?

MJacred
4:00 PM
that should resolve it for _input()
though it won't do proper clearing like InputDefault::release_pressed_events() does
but it could be good enough

lawnjelly
4:04 PM
So it needs to remove the axis from _joy_axis map I'm guessing to be complete.

orlambda committed:
Address issues

Only sends axis event if value is not 0.0
Parses events immediately without pushing to buffered_events
Removes axes from _joy_axis map

orlambda commented:
I don't know if it's ok to use `_parse_input_event_impl()`, I'd need someone with more knowledge judge that.

Yes we'd need to deal with `buffered_events`. Would looping through the list be acceptable for performance? Another approach could be to check in `_parse_input_event_impl()` whether a joypad is connected and ignore the event if not. I'm not sure how this would work with multithreading to make sure no invalid values are parsed before the `connected` is changed to `false`. Would this be a good check to add anyway?

I think it's premature to implement the above before it's clear that pushing the reset events ahead of `buffered_events` is actually safe and beneficial. It would be good to hear how this works on other systems as I don't see a benefit to this on mine.

If I understand correctly, the point of this is to eliminate a one frame delay in resetting the values. I wonder if that would be an acceptably minor issue given the contexts in which joypads would usually be disconnected?

lawnjelly commented:
> Any thoughts on what this needs to move forward?

Sorry I didn't see your original reply.

It's a while since I read the PR but my own main concern I think was the `buffered_events` on OSes that use these could still maybe contain events from the controller being deleted.

So the implication is that when removing a controller we should probably also go through `buffered_events` and remove any that concern the removed controller, so we don't later start getting events coming in from a removed controller.

As I say I'm not an input specialist so I'm having to play it safe here unless we get an approval from an input maintainer.

Another alternative might be to post the controller removed message to the `buffered_events`, and process everything in order. :thinking:
